File: code/python/exercises.py
from dataclasses import dataclass
from typing import Callable, Generator, Sequence, TypeVar

Number = int | float
T = TypeVar("T")


def first_then_apply(
    strings: Sequence[str],
    predicate: Callable[[str], bool],
    function: Callable[[str], T],
) -> T | None:
    for x in strings:
        if predicate(x):
            return function(x)
    return None


def powers_generator(*, base: Number, limit: Number) -> Generator[Number, None, None]:
    value: Number = 1
    while value <= limit:
        yield value
        value *= base


def meaningful_line_count(filename: str, /) -> int:
    count = 0
    with open(filename, "r", encoding="utf-8") as f:
        for line in f:
            s = line.lstrip()
            if s == "" or s == "\n":
                continue
            if s.startswith("#"):
                continue
            if s.strip() == "":
                continue
            count += 1
    return count


def say(word: str | None = None, /):
    words: list[str] = []

    def inner(next_word: str | None = None, /):
        nonlocal words
        if next_word is None:
            return " ".join(words)
        words.append(next_word)
        return inner

    if word is None:
        return ""
    words.append(word)
    return inner


@dataclass(frozen=True)
class Quaternion:
    a: Number = 0
    b: Number = 0
    c: Number = 0
    d: Number = 0

    def __add__(self, o: "Quaternion") -> "Quaternion":
        return Quaternion(self.a + o.a, self.b + o.b, self.c + o.c, self.d + o.d)

    def __mul__(self, o: "Quaternion") -> "Quaternion":
        a1, b1, c1, d1 = self.a, self.b, self.c, self.d
        a2, b2, c2, d2 = o.a, o.b, o.c, o.d
        return Quaternion(
            a1 * a2 - b1 * b2 - c1 * c2 - d1 * d2,
            a1 * b2 + b1 * a2 + c1 * d2 - d1 * c2,
            a1 * c2 - b1 * d2 + c1 * a2 + d1 * b2,
            a1 * d2 + b1 * c2 - c1 * b2 + d1 * a2,
        )

    @property
    def coefficients(self) -> tuple[Number, Number, Number, Number]:
        return (float(self.a), float(self.b), float(self.c), float(self.d))

    @property
    def conjugate(self) -> "Quaternion":
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __str__(self) -> str:
        parts: list[str] = []
        if self.a != 0:
            parts.append(f"{self.a}")
        for coeff, sym in ((self.b, "i"), (self.c, "j"), (self.d, "k")):
            if coeff == 0:
                continue
            sign = "+" if coeff > 0 else "-"
            mag = abs(coeff)
            term = sym if mag == 1 else f"{mag}{sym}"
            if not parts:
                parts.append(term if coeff > 0 else f"-{term}")
            else:
                parts.append(f"{sign}{term}")
        return "".join(parts) if parts else "0"

File: code/lua/exercises.lua
function first_then_apply(array, predicate, func)
    for _, s in ipairs(array) do
        if predicate(s) then
            return func(s)
        end
    end
    return nil
end

function powers_generator(base, limit)
    return coroutine.create(function()
        local value = 1
        while value <= limit do
            coroutine.yield(value)
            value = value * base
        end
    end)
end

function meaningful_line_count(filename)
    local file, err = io.open(filename, "r")
    if not file then
        error(err)
    end
    local count = 0
    for line in file:lines() do
        local trimmed = line:gsub("^%s+", "")
        if trimmed ~= "" and not trimmed:match("^%s*$") and not trimmed:match("^#") then
            count = count + 1
        end
    end
    file:close()
    return count
end

function say(word)
    if word == nil then
        return ""
    end
    local words = { word }
    local function inner(next)
        if next == nil then
            return table.concat(words, " ")
        end
        words[#words + 1] = next
        return inner
    end
    return inner
end

local QuaternionMeta = {}
QuaternionMeta.__index = QuaternionMeta

function QuaternionMeta.__add(x, y)
    return Quaternion.new(x.a + y.a, x.b + y.b, x.c + y.c, x.d + y.d)
end

function QuaternionMeta.__mul(x, y)
    local a1, b1, c1, d1 = x.a, x.b, x.c, x.d
    local a2, b2, c2, d2 = y.a, y.b, y.c, y.d
    return Quaternion.new(
        a1 * a2 - b1 * b2 - c1 * c2 - d1 * d2,
        a1 * b2 + b1 * a2 + c1 * d2 - d1 * c2,
        a1 * c2 - b1 * d2 + c1 * a2 + d1 * b2,
        a1 * d2 + b1 * c2 - c1 * b2 + d1 * a2
    )
end

function QuaternionMeta.__eq(x, y)
    return x.a == y.a and x.b == y.b and x.c == y.c and x.d == y.d
end

function QuaternionMeta.__tostring(q)
    local parts = {}
    if q.a ~= 0 then
        parts[#parts + 1] = tostring(q.a)
    end
    local function add_term(coeff, sym)
        if coeff == 0 then
            return
        end
        local sign = coeff > 0 and "+" or "-"
        local mag = math.abs(coeff)
        local body = (mag == 1) and sym or (tostring(mag) .. sym)
        if #parts == 0 then
            parts[#parts + 1] = (coeff > 0) and body or ("-" .. body)
        else
            parts[#parts + 1] = sign .. body
        end
    end
    add_term(q.b, "i")
    add_term(q.c, "j")
    add_term(q.d, "k")
    return (#parts > 0) and table.concat(parts) or "0"
end

function QuaternionMeta:coefficients()
    return { self.a, self.b, self.c, self.d }
end

function QuaternionMeta:conjugate()
    return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

Quaternion = {}

function Quaternion.new(a, b, c, d)
    local obj = {
        a = tonumber(a) or 0,
        b = tonumber(b) or 0,
        c = tonumber(c) or 0,
        d = tonumber(d) or 0
    }
    return setmetatable(obj, QuaternionMeta)
end


